File: apps/backend/src/models/deduction_profile.rs
use chrono::{NaiveTime, Timelike};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::models::parse_time;

#[derive(Debug, Clone, Serialize, Deserialize, ToSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct DeductionProfile {
    pub peak_window_start: String,
    pub peak_window_end: String,
    pub peak_ratio: f64,
    pub low_window_start: String,
    pub low_window_end: String,
    pub low_ratio: f64,
}

impl DeductionProfile {
    pub fn validate(&self) -> Result<(), String> {
        for field in [
            &self.peak_window_start,
            &self.peak_window_end,
            &self.low_window_start,
            &self.low_window_end,
        ] {
            if parse_time(field).is_err() {
                return Err(format!("Invalid time format: {field}"));
            }
        }
        if self.peak_ratio <= 1.0 {
            return Err("peakRatio must be greater than 1".to_string());
        }
        if self.low_ratio <= 0.0 || self.low_ratio >= 1.0 {
            return Err("lowRatio must be between 0 and 1 (exclusive)".to_string());
        }
        if windows_overlap(self) {
            return Err("Peak and low windows must not overlap".to_string());
        }
        Ok(())
    }
}

pub fn minute_in_window(minute_of_day: i32, start: &str, end: &str) -> bool {
    let Ok(start_t) = parse_time(start) else {
        return false;
    };
    let Ok(end_t) = parse_time(end) else {
        return false;
    };
    time_in_window(minute_of_day, start_t, end_t)
}

fn minute_of_day(time: NaiveTime) -> i32 {
    time.hour() as i32 * 60 + time.minute() as i32
}

fn time_in_window(minute: i32, start: NaiveTime, end: NaiveTime) -> bool {
    let s = minute_of_day(start);
    let e = minute_of_day(end);
    if s == e {
        return false;
    }
    if s < e {
        minute >= s && minute < e
    } else {
        minute >= s || minute < e
    }
}

pub fn ratio_at_time(local_time: NaiveTime, profile: &DeductionProfile) -> f64 {
    let minute = minute_of_day(local_time);
    if minute_in_window(minute, &profile.peak_window_start, &profile.peak_window_end) {
        return profile.peak_ratio;
    }
    if minute_in_window(minute, &profile.low_window_start, &profile.low_window_end) {
        return profile.low_ratio;
    }
    1.0
}

pub fn windows_overlap(profile: &DeductionProfile) -> bool {
    for minute in 0..(24 * 60) {
        if minute_in_window(minute, &profile.peak_window_start, &profile.peak_window_end)
            && minute_in_window(minute, &profile.low_window_start, &profile.low_window_end)
        {
            return true;
        }
    }
    false
}

File: apps/console-tv/app/src/main/java/com/gamingcafe/consoletv/domain/DeductionProfile.kt
package com.gamingcafe.consoletv.domain

import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

data class DeductionProfile(
    val peakWindowStart: String,
    val peakWindowEnd: String,
    val peakRatio: Double,
    val lowWindowStart: String,
    val lowWindowEnd: String,
    val lowRatio: Double,
)

fun parseTimeToMinutes(value: String): Int {
    val parts = value.split(":").map { it.toIntOrNull() ?: 0 }
    val hour = parts.getOrElse(0) { 0 }
    val minute = parts.getOrElse(1) { 0 }
    return hour * 60 + minute
}

fun minuteInWindow(minuteOfDay: Int, start: String, end: String): Boolean {
    val s = parseTimeToMinutes(start)
    val e = parseTimeToMinutes(end)
    if (s == e) return false
    return if (s < e) {
        minuteOfDay in s until e
    } else {
        minuteOfDay >= s || minuteOfDay < e
    }
}

fun ratioAtMinute(minuteOfDay: Int, profile: DeductionProfile): Double {
    if (minuteInWindow(minuteOfDay, profile.peakWindowStart, profile.peakWindowEnd)) {
        return profile.peakRatio
    }
    if (minuteInWindow(minuteOfDay, profile.lowWindowStart, profile.lowWindowEnd)) {
        return profile.lowRatio
    }
    return 1.0
}

fun localMinuteOfDay(instant: Instant, cafeTimezone: String): Int {
    val zoned = ZonedDateTime.ofInstant(instant, ZoneId.of(cafeTimezone))
    return zoned.hour * 60 + zoned.minute
}

fun currentDeductionRatio(
    profile: DeductionProfile,
    cafeTimezone: String,
    now: Instant = Instant.now(),
): Double = ratioAtMinute(localMinuteOfDay(now, cafeTimezone), profile)
